File: two-sum-iv-input-is-a-bst/two-sum-iv-input-is-a-bst.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def findTarget(self, root: Optional[TreeNode], k: int) -> bool:
        # x + y = k
        # y = k - x
        T = set()
        
        def dfs(node):
            if not node:
                return False
            y = k-node.val
            if y in T:
                return True
            else:
                T.add(node.val)
            return dfs(node.left) or dfs(node.right)
        
        return dfs(root)
        
        

File: two-sum-iv-input-is-a-bst/two-sum-iv-input-is-a-bst.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    set<int> l;
    bool findTarget(TreeNode* root, int k) {
        return dfs(root,k);
    }
    bool dfs(TreeNode* root, int k){
        if (!root) return false;
        int y = k - root->val;
        if (l.find(y)!= l.end()) return true;
        else {
            l.insert(root->val);
            return dfs(root->left, k) || dfs(root->right, k);
        }
    }
};